digitalocean/resource_digitalocean_kubernetes_cluster.go

```diff
@@ -352,6 +365,87 @@ func resourceDigitalOceanKubernetesClusterDelete(d *schema.ResourceData, meta in
 	return nil
 }
 
+// Import a Kubernetes cluster and its node pools into the Terraform state.
+//
+// Note: This resource cannot make use of the pass-through importer because special handling is
+// required to ensure the default node pool has the `terraform:default-node-pool` tag and to
+// import any non-default node pools associated with the cluster.
+func resourceDigitalOceanKubernetesClusterImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	client := meta.(*CombinedConfig).godoClient()
+
+	cluster, _, err := client.Kubernetes.Get(context.Background(), d.Id())
+	if err != nil {
+		return nil, err
+	}
+
+	// Check how many node pools have the required tag. The goal is ensure that one and only one node pool
+	// has the tag (i.e., the default node pool).
+	countOfNodePoolsWithTag := 0
+	refreshCluster := false
+	for _, nodePool := range cluster.NodePools {
+		for _, tag := range nodePool.Tags {
+			if tag == digitaloceanKubernetesDefaultNodePoolTag {
+				countOfNodePoolsWithTag += 1
+			}
+		}
+	}
+	if countOfNodePoolsWithTag > 1 {
+		// Multiple node pools have the tag. Stop the import and notify the user they need to manually ensure
+		// only one node pool has the tag.
+		return nil, fmt.Errorf("Multiple node pools are tagged as the default node pool; only one node pool may have the `%s` tag", digitaloceanKubernetesDefaultNodePoolTag)
+	} else if countOfNodePoolsWithTag == 0 {
+		// None of the node pools have the tag. If there is only one node pool, then it must be the default
+		// node pool and thus add the tag. Adding the tag is non-destructive, and thus should be fine.
+		if len(cluster.NodePools) == 1 {
+			nodePool := cluster.NodePools[0]
+			tags := append(nodePool.Tags, digitaloceanKubernetesDefaultNodePoolTag)
+
+			nodePoolUpdateRequest := &godo.KubernetesNodePoolUpdateRequest{
+				Tags: tags,
+			}
+
+			log.Printf("[INFO] Adding %s tag to node pool %s in cluster %s", digitaloceanKubernetesDefaultNodePoolTag,
+				nodePool.ID, cluster.ID)
+
+			_, _, err := client.Kubernetes.UpdateNodePool(context.Background(), cluster.ID, nodePool.ID, nodePoolUpdateRequest)
+			if err != nil {
+				return nil, err
+			}
+
+			refreshCluster = true
+		} else {
+			return nil, fmt.Errorf("Cannot infer default node pool since there are multiple node pools; please manually add the `%s` tag to the default node pool", digitaloceanKubernetesDefaultNodePoolTag)
+		}
+	}
+
+	// Refresh the cluster and node pools so that any the added tag
+	if refreshCluster {
+		cluster, _, err = client.Kubernetes.Get(context.Background(), d.Id())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Generate a list of ResourceData for the cluster and node pools.
+	resourceDatas := make([]*schema.ResourceData, 1)
+	resourceDatas[0] = d // the cluster
+	for _, nodePool := range cluster.NodePools {
+		for _, tag := range nodePool.Tags {
+			if tag != digitaloceanKubernetesDefaultNodePoolTag {
```

Comment:
DigitalOcean adds a number of tags to node pools automatically (e.g. `k8s`, `k8s:<cluster-uuid>`), so this loop will still attempt to add a node pool that has `digitaloceanKubernetesDefaultNodePoolTag` and import the same pool multiple times. This is why the error you noted above shows seems to be trying to import six pools.

It is looping through each tag individually and then adding the pool to the list based on that individual tag. Instead, we need to evaluate this on whether or not the tag is in the list of tags. (This is one of those times where I wish Go had a `if "x" in foo:` constructor like Python.)

There are a number of ways to do this, but this works:

	for _, nodePool := range cluster.NodePools {
		importPool := true
		for _, tag := range nodePool.Tags {
			if tag == digitaloceanKubernetesDefaultNodePoolTag {
				importPool = false
			}
		}
		if importPool {
			// Assign the resourceData
		}
	}

Reply:
Will fix.